### PR TITLE
CONSOLE-3166, Bug 2116973: Refactor nav components and console app nav extensions

### PR DESCRIPTION
**Pull request by TheRealJon**
- Replace all console app nav extension `resource` properties with appropriate `model` + `startsWith` properties
- Add a deprecated message to all static nav extension types and remove references to legacy nav components
- Update `connectToModel`, `connectToPlural`, and `Firehose` so that they resolve k8s models for core API `group~version~kind` references
- Refactor `NavLink*` class components to `NavItem*` function components to remove inheritance pattern
- Remove `NavLinkRoot` component and include logic in refactored `NavItem*` components
- Refactor `PluginNavItems` to `PluginNavItem` and map extensions from the parent
- General naming improvements
- Improve nav-related util function naming and logic

### CODE REVIEW

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Everything looks good to me. I made one comment that ideally we do before it merges, but it really doesn't need to hold anything up.

I'm kinda shocked our extensions were so wrong, thanks for cleaning those up 👍

File: frontend/packages/console-plugin-sdk/src/typings/nav-items.ts
Comment (TheRealJon) [reply frontend/packages/console-plugin-sdk/src/typings/nav-items.ts:25]: Yeah, we have a separate story for deleting these files, but I had to update the types since they were referencing components that I changed. I didn't want to make updates here without also clarifying that these extensions are not actually supported.

**Comment by TheRealJon** [comment]
> I'm kinda shocked our extensions were so wrong, thanks for cleaning those up 👍

Yeah, that was leftover tech debt from one of my previous stories. Things got ugly 😆

File: frontend/packages/console-plugin-sdk/src/typings/nav-items.ts
Comment (TheRealJon) [reply frontend/packages/console-plugin-sdk/src/typings/nav-items.ts:25]: In reality, these changes might not even end up in master because @zherman0 is working on the story to delete this. We could merge that first, which would make these changes obsolete.

**Review by andrewballantyne — APPROVED** [review]
This is good to go from my perspective. @TheRealJon not sure if you're looking to merge this in a specific order, you alluded to potentially waiting on other work... but I'm fine with this as long as we delete those deprecated items in a timely manner -- no sense in adding to our technical debt.

**Comment by andrewballantyne** [comment]
I was doing some edge case testing and I'm not sure if I am just blundering my configuration or not -- but I think `startsWith` isn't doing what it's supposed to be doing 🤔 

```
diff --git a/dynamic-demo-plugin/console-extensions.json b/dynamic-demo-plugin/console-extensions.json
index 57ce393b66..3bc3756d16 100644
--- a/dynamic-demo-plugin/console-extensions.json
+++ b/dynamic-demo-plugin/console-extensions.json
@@ -89,6 +89,7 @@
       "perspective": "admin",
       "section": "admin-demo-section",
       "name": "%plugin__console-demo-plugin~Dynamic Nav 1%",
+      "startsWith": ["dynamic-route"],
       "href": "/dynamic-route-1"
     }
   },
```
This doesn't work... I'd expect both dynamic-route hrefs to highlight when I click on and navigate to `/dynamic-route-2` -- if I make it `"startsWith": ["dynamic-route-2"],` it works, but that seems too specific.

File: frontend/packages/console-app/console-extensions.json
Comment (andrewballantyne) [inline frontend/packages/console-app/console-extensions.json:475]: The more I think about this -- especially given my other note I just made on how I thought this field worked... I am hesitant to agree this is needed. I'd think we should be able to infer this value from the model 🤔 eg a `ProjectsModel.plural` is all this is... and you describe the model effectively (with this change).

Can you look into removing the `startsWith` as much as possible? I think it's an "extra" feature for flexibility of "sub pages" and not intended to be a mainstay of the way the navigation works.

Comment (TheRealJon) [reply frontend/packages/console-app/console-extensions.json:475]: Could we address this as a follow-on tech debt issue? I only added these startsWith props so that we retain active nav items for old links, and I'm not sure it's worth the work of implementing a special case to match on model.plural.

Comment (andrewballantyne) [reply frontend/packages/console-app/console-extensions.json:475]: Personally I'd deal with it where you introduced it, but if you want to push it for some reason, it should remain a priority to fix it and not slip through the cracks. You/OCP team can determine how you want to split up the work, I'm just noting my concerns 🙂 

I view this as an issue with the extension if we are not effectively using it properly when we consume the data. `startsWith` is optional... if you do not work without it, I'd call this a bug -- other plugins will struggle through this pain too.

**Comment by TheRealJon** [comment]
> I was doing some edge case testing and I'm not sure if I am just blundering my configuration or not -- but I think `startsWith` isn't doing what it's supposed to be doing 🤔
> 
> ```
> diff --git a/dynamic-demo-plugin/console-extensions.json b/dynamic-demo-plugin/console-extensions.json
> index 57ce393b66..3bc3756d16 100644
> --- a/dynamic-demo-plugin/console-extensions.json
> +++ b/dynamic-demo-plugin/console-extensions.json
> @@ -89,6 +89,7 @@
>        "perspective": "admin",
>        "section": "admin-demo-section",
>        "name": "%plugin__console-demo-plugin~Dynamic Nav 1%",
> +      "startsWith": ["dynamic-route"],
>        "href": "/dynamic-route-1"
>      }
>    },
> ```
> 
> This doesn't work... I'd expect both dynamic-route hrefs to highlight when I click on and navigate to `/dynamic-route-2` -- if I make it `"startsWith": ["dynamic-route-2"],` it works, but that seems too specific.

The API docs state the following on this property:
> Mark this item as active when the URL starts with one of these paths.


I interpreted this to mean that it would match on entire path segments and not partial strings.

**Comment by andrewballantyne** [comment]
> The API docs state the following on this property:
> 
> > Mark this item as active when the URL starts with one of these paths.
> 
> I interpreted this to mean that it would match on entire path segments and not partial strings.

That being said -- if I recall correctly, it worked before your PR and now does not. Bug caused or bug fixed? Up to debate I guess. I find it less likely to be a bug fixed as its a loss of functionality for no specific reason -- it's ambiguous at best I feel. 

If you want to claim this is the way it _should_ worked, I'd make the definition behind the property more specific as to remove the ambiguity. `string.startsWith()` ([MDN ref](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith)) doesn't assume anything other than it starts with the string. So baring more specific information, I'd expect similar functionality as a developer.

As long as it is documented accurately, we can claim not-a-bug if someone has an issue with it and mull over the enhancement as needed.

**Comment by TheRealJon** [comment]
@andrewballantyne I just pushed up some changes that address your comments.

**Comment by andrewballantyne** [comment]
/retest
/lgtm

Thanks for the changes. Looks good.

**Comment by TheRealJon** [comment]
/retest

Cluster install failed

**Comment by jcaianirh** [comment]
QE Approver:
/assign @yapei

Docs Approver:
/assign @opayne1

PX Approver:
/assign @RickJWagner

**Comment by TheRealJon** [comment]
/hold

Found a legit CI failure

File: frontend/packages/console-app/console-extensions.json
Comment (andrewballantyne) [inline frontend/packages/console-app/console-extensions.json:518]: Is this needed? Shouldn't by the nature of having a `href` cover this?

File: frontend/public/components/nav/PluginNavItem.tsx
Comment (andrewballantyne) [inline frontend/public/components/nav/PluginNavItem.tsx:54]: Curious, if it makes it here, is this a valid state or did we want to also `console.warn` or something that we didn't render an unknown nav item?

**Review by andrewballantyne — COMMENTED** [review]
Looking good... testing around a bit.

Got a couple questions...

**Comment by yapei** [comment]
performed a regression testing about nav-related functionalities, for example, nav order, nav linking, nav state, no issues found
/label qe-approved

File: frontend/packages/console-app/console-extensions.json
Comment (TheRealJon) [reply frontend/packages/console-app/console-extensions.json:518]: This is probably not strictly necessary. It might have just been a left-over prop from the original admin nav component.

File: frontend/public/components/nav/PluginNavItem.tsx
Comment (TheRealJon) [reply frontend/public/components/nav/PluginNavItem.tsx:54]: Since this is an used component and we use type guards to make sure we are only passing nav extensions to this component, we probably don't need a warning here. We theoretically shouldn't ever reach this line.

Comment (andrewballantyne) [reply frontend/public/components/nav/PluginNavItem.tsx:54]: More of a reason I was thinking of a warning -- React does warnings for components in the render thread by having a flag they switch when it's been presented once. I was thinking we could do that.

But we can ignore this idea for now. Less of a jump on any bug that is related in the future, but probably not by much.

**Pull request merged**
